File: claude-code.js
/**
 * Claude Code CLI 交互层
 *
 * 核心设计：
 * - 每个微信用户一个独立 session（对话连续）
 * - 通过 stream-json 实时解析 Claude 的动作（读文件、写文件、执行命令）
 * - 通过 onProgress 回调实时反馈给用户
 * - 使用 --dangerously-skip-permissions 赋予完整的代码操作能力
 */

import { spawn, execFileSync } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';

// ── 状态管理 ─────────────────────────────────────────────────────────────────

const sessions = new Map();   // userId → { sessionId, lastActive }
const locks = new Map();      // userId → Promise（同用户串行）

let activeProcesses = 0;
const MAX_CONCURRENT = 3;
const MAX_SESSIONS = 100;
const SESSION_EXPIRE_MS = 60 * 60 * 1000;

const activeChildren = new Set();

// ── 工具名称友好映射 ─────────────────────────────────────────────────────────

const TOOL_LABELS = {
  Read: '📖 正在读取文件',
  Edit: '✏️ 正在编辑文件',
  Write: '📝 正在创建文件',
  Bash: '⚡ 正在执行命令',
  Glob: '🔍 正在搜索文件',
  Grep: '🔍 正在搜索内容',
  WebSearch: '🌐 正在搜索网页',
  WebFetch: '🌐 正在获取网页',
  TodoWrite: '📋 正在规划任务',
};

/**
 * 从 tool_use 事件中提取友好的进度描述
 */
function describeToolUse(toolName, input) {
  const label = TOOL_LABELS[toolName] || `🔧 ${toolName}`;

  if (!input) return label;

  switch (toolName) {
    case 'Read':
      return `${label}: ${extractFilename(input.file_path)}`;
    case 'Edit':
      return `${label}: ${extractFilename(input.file_path)}`;
    case 'Write':
      return `${label}: ${extractFilename(input.file_path)}`;
    case 'Bash':
      return `${label}: ${truncate(input.command || input.description || '', 60)}`;
    case 'Glob':
      return `${label}: ${input.pattern || ''}`;
    case 'Grep':
      return `${label}: ${truncate(input.pattern || '', 40)}`;
    default:
      return label;
  }
}

function extractFilename(filepath) {
  if (!filepath) return '';
  const parts = filepath.replace(/\\/g, '/').split('/');
  return parts.slice(-2).join('/');
}

function truncate(str, maxLen) {
  if (str.length <= maxLen) return str;
  return str.slice(0, maxLen) + '...';
}

// ── 主要 API ─────────────────────────────────────────────────────────────────

/**
 * 发送消息给 Claude Code
 * @param {string} userId - 微信用户ID
 * @param {string} message - 用户消息
 * @param {object} opts
 * @param {string} opts.cwd - 工作目录
 * @param {function} opts.onProgress - 进度回调 (progressText: string) => void
 * @returns {Promise<string>} Claude 的回复
 */
export async function chat(userId, message, opts = {}) {
  // 同用户串行
  while (locks.get(userId)) {
    await locks.get(userId);
  }

  // 全局并发等待（静默排队，不发消息打扰用户）
  while (activeProcesses >= MAX_CONCURRENT) {
    await new Promise(r => setTimeout(r, 1000));
  }

  let resolveLock;
  const lockPromise = new Promise(r => { resolveLock = r; });
  locks.set(userId, lockPromise);

  try {
    return await _doChat(userId, message, opts);
  } finally {
    locks.delete(userId);
    resolveLock();
  }
}

async function _doChat(userId, message, opts) {
  const session = sessions.get(userId);
  const now = Date.now();

  cleanupExpiredSessions();

  const isExpired = session && (now - session.lastActive > SESSION_EXPIRE_MS);
  const sessionId = (!isExpired && session?.sessionId) || null;

  const args = [
    '-p',
    '--output-format', 'stream-json',
    // 关键：赋予完整权限，让 Claude Code 能真正写代码、执行命令
    '--dangerously-skip-permissions',
  ];

  if (sessionId) {
    args.push('-r', sessionId);
  }

  if (opts.model) {
    args.push('--model', opts.model);
  }

  if (opts.allowedTools?.length) {
    args.push('--allowedTools', opts.allowedTools.join(','));
  }

  // 消息作为最后一个位置参数
  args.push(message);

  const { bin, extraArgs, shell } = resolveClaudeSpawn();
  const onProgress = opts.onProgress || (() => {});

  return new Promise((resolve, reject) => {
    activeProcesses++;

    const proc = spawn(bin, [...extraArgs, ...args], {
      stdio: ['pipe', 'pipe', 'pipe'],
      shell,
      cwd: opts.cwd || undefined,
      env: { ...process.env, CLAUDECODE: undefined },
    });

    activeChildren.add(proc);

    let stdout = '';
    let stderr = '';
    let assistantText = '';
    let finalResult = '';
    let newSessionId = null;
    let toolUseCount = 0;
    let lastProgressTime = 0;
    const PROGRESS_THROTTLE_MS = 3000; // 进度消息最少间隔 3 秒

    proc.stdout.on('data', (chunk) => {
      stdout += chunk.toString();
      const lines = stdout.split('\n');
      stdout = lines.pop() || '';

      for (const line of lines) {
        if (!line.trim()) continue;
        try {
          handleEvent(JSON.parse(line));
        } catch {}
      }
    });

    proc.stderr.on('data', (chunk) => {
      stderr += chunk.toString();
    });

    function handleEvent(event) {
      switch (event.type) {
        case 'assistant': {
          if (event.session_id) newSessionId = event.session_id;
          if (!event.message?.content) break;

          for (const block of event.message.content) {
            if (block.type === 'text') {
              assistantText += block.text;
            }
            // 关键：解析 tool_use 事件，实时反馈给用户
            if (block.type === 'tool_use') {
              toolUseCount++;
              const now = Date.now();
              if (now - lastProgressTime >= PROGRESS_THROTTLE_MS) {
                const desc = describeToolUse(block.name, block.input);
                onProgress(desc);
                lastProgressTime = now;
              }
            }
          }
          break;
        }
        case 'result': {
          if (event.session_id) newSessionId = event.session_id;
          if (event.result && typeof event.result === 'string') {
            finalResult = event.result;
          }
          break;
        }
      }
    }

    function getReplyText() {
      return finalResult || assistantText;
    }

    function cleanup() {
      activeProcesses = Math.max(0, activeProcesses - 1);
      activeChildren.delete(proc);
    }

    proc.on('close', (code) => {
      cleanup();

      if (stdout.trim()) {
        try { handleEvent(JSON.parse(stdout)); } catch {}
      }

      if (code !== 0 && !getReplyText()) {
        if (sessionId && (stderr.includes('session') || stderr.includes('conversation'))) {
          sessions.delete(userId);
          _doChat(userId, message, opts).then(resolve).catch(reject);
          return;
        }
        reject(new Error(`Claude Code 退出码 ${code}: ${stderr.slice(0, 500)}`));
        return;
      }

      // 更新 session
      if (newSessionId) {
        sessions.set(userId, { sessionId: newSessionId, lastActive: Date.now() });
      } else if (session) {
        session.lastActive = Date.now();
      }

      resolve(getReplyText() || '(Claude Code 无响应)');
    });

    proc.on('error', (err) => {
      cleanup();
      reject(new Error(`无法启动 claude 命令: ${err.message}\n请确认已安装: npm install -g @anthropic-ai/claude-code`));
    });

    // 5分钟硬超时，不再发烦人的"仍在工作中"提醒
    const killTimer = setTimeout(() => {
      try { proc.kill('SIGTERM'); } catch {}
      setTimeout(() => { try { proc.kill('SIGKILL'); } catch {} }, 5000);
    }, 5 * 60_000);

    proc.on('close', () => clearTimeout(killTimer));
  });
}

// ── 辅助函数 ─────────────────────────────────────────────────────────────────

function cleanupExpiredSessions() {
  const now = Date.now();
  for (const [userId, session] of sessions) {
    if (now - session.lastActive > SESSION_EXPIRE_MS) {
      sessions.delete(userId);
    }
  }
  if (sessions.size > MAX_SESSIONS) {
    const sorted = [...sessions.entries()].sort((a, b) => a[1].lastActive - b[1].lastActive);
    for (const [userId] of sorted.slice(0, sessions.size - MAX_SESSIONS)) {
      sessions.delete(userId);
    }
  }
}

export function clearSession(userId) {
  sessions.delete(userId);
}

export function clearAllSessions() {
  sessions.clear();
}

export function killAll() {
  for (const proc of activeChildren) {
    try { proc.kill('SIGTERM'); } catch {}
  }
}

// ── 平台适配 ─────────────────────────────────────────────────────────────────

const IS_WINDOWS = process.platform === 'win32';

/**
 * 解析 Claude Code 的启动方式
 *
 * Windows 上 claude 是 .cmd 文件，通过 cmd.exe 执行会破坏中文参数编码。
 * 解决方案：找到 cli.js 入口，直接用 node 调用，完全绕过 cmd.exe。
 *
 * 返回 { bin, extraArgs, shell }：
 *   Windows: { bin: 'node', extraArgs: ['/.../cli.js'], shell: false }
 *   其他:    { bin: 'claude', extraArgs: [],              shell: false }
 */
let _spawnCache = null;
function resolveClaudeSpawn() {
  if (_spawnCache) return _spawnCache;

  if (IS_WINDOWS) {
    try {
      // 找到 claude.cmd 路径
      const result = execFileSync('where', ['claude'], {
        encoding: 'utf-8',
        shell: true,
        env: { ...process.env, CLAUDECODE: undefined },
      }).trim();
      const cmdPath = result.split('\n').map(l => l.trim()).find(l => l.endsWith('.cmd'));
      if (cmdPath) {
        // claude.cmd 和 node_modules 在同一目录
        const dir = path.dirname(cmdPath);
        const cliJs = path.join(dir, 'node_modules', '@anthropic-ai', 'claude-code', 'cli.js');
        if (fs.existsSync(cliJs)) {
          _spawnCache = { bin: process.execPath, extraArgs: [cliJs], shell: false };
          return _spawnCache;
        }
      }
    } catch {}
    // 兜底：仍用 cmd.exe（可能中文有问题，但至少能跑）
    _spawnCache = { bin: 'claude', extraArgs: [], shell: true };
    return _spawnCache;
  }

  // macOS / Linux：直接用 claude 命令
  try {
    const result = execFileSync('which', ['claude'], {
      encoding: 'utf-8',
      env: { ...process.env, CLAUDECODE: undefined },
    }).trim();
    _spawnCache = { bin: result || 'claude', extraArgs: [], shell: false };
  } catch {
    _spawnCache = { bin: 'claude', extraArgs: [], shell: false };
  }
  return _spawnCache;
}

export async function checkClaudeAvailable() {
  try {
    const { bin, extraArgs, shell } = resolveClaudeSpawn();
    const result = execFileSync(bin, [...extraArgs, '--version'], {
      encoding: 'utf-8',
      shell,
      env: { ...process.env, CLAUDECODE: undefined },
    });
    return result.trim();
  } catch {
    return null;
  }
}

File: adapters/claude_code.py
"""
Claude Code CLI 适配器
通过 stream-json 模式与本机 Claude Code CLI 交互
"""

import json
import os
import subprocess
import threading
import time
import logging
from pathlib import Path

from .base import ModelAdapter, ChatResult
from core.config import STATE_DIR, PROCESS_TIMEOUT

log = logging.getLogger(__name__)

# ── 工具标签映射 ─────────────────────────────────────────────────────────────

TOOL_LABELS = {
    'Read': '📖 正在读取文件',
    'Edit': '✏️ 正在编辑文件',
    'Write': '📝 正在创建文件',
    'Bash': '⚡ 正在执行命令',
    'Glob': '🔍 正在搜索文件',
    'Grep': '🔍 正在搜索内容',
    'WebSearch': '🌐 正在搜索网页',
    'WebFetch': '🌐 正在获取网页',
    'TodoWrite': '📋 正在规划任务',
}

SESSION_FILE = STATE_DIR / 'sessions.json'


def _describe_tool_use(tool_name: str, tool_input: dict | None) -> str:
    label = TOOL_LABELS.get(tool_name, f'🔧 {tool_name}')
    if not tool_input:
        return label
    if tool_name in ('Read', 'Edit', 'Write'):
        fp = tool_input.get('file_path', '')
        parts = fp.replace('\\', '/').split('/')
        return f'{label}: {"/".join(parts[-2:])}'
    if tool_name == 'Bash':
        cmd = tool_input.get('command') or tool_input.get('description') or ''
        return f'{label}: {cmd[:60]}{"..." if len(cmd) > 60 else ""}'
    if tool_name == 'Glob':
        return f'{label}: {tool_input.get("pattern", "")}'
    if tool_name == 'Grep':
        pat = tool_input.get('pattern', '')
        return f'{label}: {pat[:40]}{"..." if len(pat) > 40 else ""}'
    return label


# ── 平台适配 ─────────────────────────────────────────────────────────────────

IS_WINDOWS = os.name == 'nt'
_spawn_cache = None


def _resolve_claude_spawn() -> dict:
    """解析 Claude Code CLI 启动方式"""
    global _spawn_cache
    if _spawn_cache:
        return _spawn_cache

    if IS_WINDOWS:
        try:
            result = subprocess.run(
                ['where', 'claude.cmd'],
                capture_output=True, text=True, shell=True, timeout=5,
            )
            cmd_path = result.stdout.strip().splitlines()[0].strip()
            if cmd_path:
                dir_path = Path(cmd_path).parent
                cli_js = dir_path / 'node_modules' / '@anthropic-ai' / 'claude-code' / 'cli.js'
                if cli_js.exists():
                    import sys
                    # 用 node 直接调用 cli.js，绕过 cmd.exe 编码问题
                    node_path = subprocess.run(
                        ['where', 'node'], capture_output=True, text=True, shell=True, timeout=5
                    ).stdout.strip().splitlines()[0].strip()
                    _spawn_cache = {'bin': node_path, 'extra_args': [str(cli_js)], 'shell': False}
                    return _spawn_cache
        except Exception:
            pass
        _spawn_cache = {'bin': 'claude', 'extra_args': [], 'shell': True}
        return _spawn_cache

    # macOS / Linux
    try:
        result = subprocess.run(['which', 'claude'], capture_output=True, text=True, timeout=5)
        bin_path = result.stdout.strip() or 'claude'
    except Exception:
        bin_path = 'claude'
    _spawn_cache = {'bin': bin_path, 'extra_args': [], 'shell': False}
    return _spawn_cache


def _clean_env() -> dict:
    """构造干净的环境变量"""
    env = os.environ.copy()
    env.pop('CLAUDECODE', None)
    return env


# ── Claude Code 适配器 ──────────────────────────────────────────────────────

class ClaudeCodeAdapter(ModelAdapter):

    def __init__(self):
        self._sessions: dict = self._load_sessions()
        self._locks: dict[str, threading.Lock] = {}
        self._active_procs: set = set()
        self._active_count = 0
        self._count_lock = threading.Lock()
        self._max_concurrent = 3
        self._max_sessions = 100

    @property
    def name(self) -> str:
        return 'claude-code'

    # ── 会话管理 ────────────────────────────────────────────────────────

    def _load_sessions(self) -> dict:
        try:
            return json.loads(SESSION_FILE.read_text(encoding='utf-8'))
        except (FileNotFoundError, json.JSONDecodeError):
            return {}

    def _save_sessions(self):
        try:
            STATE_DIR.mkdir(exist_ok=True)
            SESSION_FILE.write_text(json.dumps(self._sessions), encoding='utf-8')
        except Exception:
            pass

    def clear_session(self, user_id: str):
        self._sessions.pop(user_id, None)
        self._save_sessions()

    def _cleanup_oldest(self):
        if len(self._sessions) <= self._max_sessions:
            return
        sorted_items = sorted(self._sessions.items(), key=lambda x: x[1].get('lastActive', 0))
        for uid, _ in sorted_items[:len(self._sessions) - self._max_sessions]:
            del self._sessions[uid]
        self._save_sessions()

    # ── 核心 API ────────────────────────────────────────────────────────

    def check_available(self) -> str | None:
        try:
            spawn = _resolve_claude_spawn()
            result = subprocess.run(
                [spawn['bin'], *spawn['extra_args'], '--version'],
                capture_output=True, text=True,
                shell=spawn['shell'],
                env=_clean_env(),
                timeout=10,
            )
            return result.stdout.strip() or None
        except Exception:
            return None

    def chat(self, user_id: str, message: str, *,
             cwd: str | None = None,
             model: str | None = None,
             system_prompt: str | None = None,
             on_progress=None) -> ChatResult:

        # 同用户串行
        lock = self._locks.setdefault(user_id, threading.Lock())
        lock.acquire()
        try:
            # 全局并发等待
            while True:
                with self._count_lock:
                    if self._active_count < self._max_concurrent:
                        self._active_count += 1
                        break
                time.sleep(1)

            return self._do_chat(user_id, message, cwd=cwd, model=model,
                                 system_prompt=system_prompt, on_progress=on_progress)
        finally:
            with self._count_lock:
                self._active_count = max(0, self._active_count - 1)
            lock.release()

    def _do_chat(self, user_id: str, message: str, *,
                 cwd: str | None, model: str | None,
                 system_prompt: str | None, on_progress,
                 retry_count: int = 0) -> ChatResult:

        self._cleanup_oldest()
        session_id = (self._sessions.get(user_id) or {}).get('sessionId')

        spawn = _resolve_claude_spawn()
        args = [
            spawn['bin'], *spawn['extra_args'],
            '-p',
            '--output-format', 'stream-json',
            '--verbose',
            '--dangerously-skip-permissions',
        ]

        if session_id:
            args.extend(['-r', session_id])
        if model:
            args.extend(['--model', model])
        if system_prompt:
            args.extend(['--append-system-prompt', system_prompt])
        args.append(message)

        proc = subprocess.Popen(
            args,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            stdin=subprocess.DEVNULL,
            shell=spawn['shell'],
            cwd=cwd or None,
            env=_clean_env(),
            creationflags=subprocess.CREATE_NO_WINDOW if IS_WINDOWS else 0,
        )
        self._active_procs.add(proc)

        assistant_text = ''
        final_result = ''
        new_session_id = None
        written_files = []
        read_media_files = []
        last_progress_time = 0
        first_event = False
        timed_out = False

        # 3 秒无响应发"思考中"
        thinking_sent = False

        def read_stdout():
            nonlocal assistant_text, final_result, new_session_id
            nonlocal last_progress_time, first_event, thinking_sent

            buf = ''
            for raw_line in iter(proc.stdout.readline, b''):
                line = raw_line.decode('utf-8', errors='replace').strip()
                if not line:
                    continue

                if not first_event:
                    first_event = True

                try:
                    event = json.loads(line)
                except json.JSONDecodeError:
                    continue

                if event.get('type') == 'assistant':
                    if event.get('session_id'):
                        new_session_id = event['session_id']
                    for block in (event.get('message') or {}).get('content', []):
                        if block.get('type') == 'text':
                            assistant_text += block['text']
                        elif block.get('type') == 'tool_use':
                            inp = block.get('input') or {}
                            # 记录文件操作
                            if block.get('name') == 'Write' and inp.get('file_path'):
                                written_files.append(inp['file_path'])
                            if block.get('name') == 'Read' and inp.get('file_path'):
                                ext = Path(inp['file_path']).suffix.lower()
                                media_exts = {'.png', '.jpg', '.jpeg', '.gif', '.webp', '.bmp',
                                              '.mp4', '.mov', '.pdf', '.doc', '.docx',
                                              '.xls', '.xlsx', '.csv', '.zip', '.mp3', '.wav'}
                                if ext in media_exts:
                                    read_media_files.append(inp['file_path'])
                            # 进度回调
                            now = time.time()
                            if on_progress and now - last_progress_time >= 3:
                                desc = _describe_tool_use(block['name'], inp)
                                on_progress(desc)
                                last_progress_time = now

                elif event.get('type') == 'result':
                    if event.get('session_id'):
                        new_session_id = event['session_id']
                    if event.get('result') and isinstance(event['result'], str):
                        final_result = event['result']

        reader_thread = threading.Thread(target=read_stdout, daemon=True)
        reader_thread.start()

        # 3 秒思考提示
        def thinking_timer():
            nonlocal thinking_sent
            time.sleep(3)
            if not first_event and on_progress:
                on_progress('🧠 正在思考...')
                thinking_sent = True

        thinking_t = threading.Thread(target=thinking_timer, daemon=True)
        thinking_t.start()

        # 等待完成（带超时）
        try:
            proc.wait(timeout=PROCESS_TIMEOUT)
        except subprocess.TimeoutExpired:
            timed_out = True
            proc.terminate()
            try:
                proc.wait(timeout=5)
            except subprocess.TimeoutExpired:
                proc.kill()

        reader_thread.join(timeout=5)
        self._active_procs.discard(proc)

        stderr = proc.stderr.read().decode('utf-8', errors='replace') if proc.stderr else ''
        reply = final_result or assistant_text

        if timed_out:
            return ChatResult(
                text=reply or '⏱️ Claude Code 处理超时（5分钟），请拆分成更小的步骤。',
                written_files=written_files,
                read_media_files=read_media_files,
            )

        if proc.returncode != 0 and not reply:
            if retry_count < 1 and session_id and ('session' in stderr or 'conversation' in stderr):
                self._sessions.pop(user_id, None)
                return self._do_chat(user_id, message, cwd=cwd, model=model,
                                     system_prompt=system_prompt, on_progress=on_progress,
                                     retry_count=retry_count + 1)
            raise RuntimeError(f'Claude Code 退出码 {proc.returncode}: {stderr[:500]}')

        # 更新会话
        if new_session_id:
            self._sessions[user_id] = {'sessionId': new_session_id, 'lastActive': time.time()}
            self._save_sessions()
        elif user_id in self._sessions:
            self._sessions[user_id]['lastActive'] = time.time()

        return ChatResult(
            text=reply or '(Claude Code 无响应)',
            written_files=written_files,
            read_media_files=read_media_files,
        )

    def kill_all(self):
        for proc in list(self._active_procs):
            try:
                proc.terminate()
            except Exception:
                pass
        self._active_procs.clear()
